src/pybind/ceph_volume_client.py

```diff
@@ -0,0 +1,690 @@
+        # # the rest of the volume already exists too.
+        # try:
+        #     auth = self._rados_command("auth get", {"entity": client_entity})[0]
+        # except rados.Error:
+        #     pass
+        # else:
+        #     log.info("create_volume: {0} already exists, returning")
+        #     return {
+        #         'volume_key': auth['key'],
+        #         'mount_path': path
+        #     }
+
+        self._mkdir_p(path)
+
+        if size is not None:
+            self.fs.setxattr(path, 'ceph.quota.max_bytes', size.__str__(), 0)
+
+        if data_isolated:
+            pool_name = "{0}{1}".format(self.POOL_PREFIX, volume_path.volume_id)
+            pool_id = self._create_volume_pool(pool_name)
+            mds_map = self._rados_command("mds dump", {})
+            if pool_id not in mds_map['data_pools']:
+                self._rados_command("mds add_data_pool", {
+                    'pool': pool_name
+                })
+            self.fs.setxattr(path, 'ceph.dir.layout.pool', pool_name, 0)
+
+        return {
+            'mount_path': path
+        }
+
+    def delete_volume(self, volume_path, data_isolated=False):
+        """
+        Remove all trace of a volume from the Ceph cluster.  This function is
+        idempotent.
+
+        :param volume_path: Same identifier used in create_volume
+        :return:
+        """
+
+        log.info("delete_volume: {0}".format(volume_path))
+
+        # Create the trash folder if it doesn't already exist
+        trash = os.path.join(self.VOLUME_PREFIX, "_deleting")
+        self._mkdir_p(trash)
+
+        # We'll move it to here
+        trashed_volume = os.path.join(trash, volume_path.volume_id)
+
+        # Move the volume's data to the trash folder
+        path = self._get_path(volume_path)
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            log.warning("Trying to delete volume '{0}' but it's already gone".format(
+                path))
+        else:
+            self.fs.rename(path, trashed_volume)
+
+        try:
+            self.fs.stat(trashed_volume)
+        except cephfs.ObjectNotFound:
+            log.warning("Trying to purge volume '{0}' but it's already been purged".format(
+                trashed_volume))
+            return
+
+        def rmtree(root_path):
+            log.debug("rmtree {0}".format(root_path))
+            dir_handle = self.fs.opendir(root_path)
+            d = self.fs.readdir(dir_handle)
+            while d:
+                if d.d_name not in [".", ".."]:
+                    d_full = os.path.join(root_path, d.d_name)
+                    if d.is_dir():
+                        rmtree(d_full)
+                    else:
+                        self.fs.unlink(d_full)
+
+                d = self.fs.readdir(dir_handle)
+            self.fs.closedir(dir_handle)
+
+            self.fs.rmdir(root_path)
+
+        rmtree(trashed_volume)
```

Comment:
Why are we moving to a trash folder if we're deleting unconditionally? I imagine it's to prevent new users from mounting — but shouldn't we blacklist the clients first as well to prevent them from accessing the RADOS data with stale caps?